### ISSUE TITLE
Postmessage API - transferring files to and from

### ISSUE THREAD

**Issue by titlestad** [photopea/photopea#436]
Hello! I'm working on FriendUP - a virtual cloud computer with a full desktop environment written in JavaScript. We're working on making Photopea work, integrated into our desktop environment. We have tried to use the GET based API, but it has limited capabilities.
How do we go about using the postMessage API instead? We want to load and save files from our interface, and we'd love to be able to transfer files without having to upload them on a public server.
So f.ex. we'd like to:
* postMessage( { method: 'load', data: arraybuffer, filename: fn, type: 'jpeg' } );
* postMessage( { method: 'save', callback: id } )
  ...
  onMessage( msg )... { if( callback == id ) { process arraybuffer }

**Comment by photopea** [comment]
Hi Hogne,

if you have your files already on some server, it would be better to give Photopea the URL, so it can load it directly (through the current API). But if files are already in the users computer, passing data through postMessage would be better.

Photopea already supports scripting. What about passing data inside a script? Instead of URL, you would use data URI / base64 format (i.e. you would send a 100 MB string). 

BTW. the format ("jpeg") can be deduced from the raw bytes of the file, you don't have to pass it.

**Comment by titlestad** [comment]
I looked at your docs, but I couldn't see how this was possible. I was looking through the Adobe API to see if they supported base64 etc, but couldn't find a mention. Essentially, the API for transferring an image wasn't described anywhere. Perhaps I looked in the wrong place or didn't understand what I was reading? If so, how would you go about doing it / reading it?

**Comment by photopea** [comment]
BTW. do you know anything about the "Blob API"? I think it allows you to generate "local URLs", which are accessible only at the clients device. If you pass such URL to the current "GET API", it should work.

**Comment by titlestad** [comment]
Yes I know about it. But GET has a maximum string length, so you could possibly send a 32x32 px bitmap, but nothing of any useful size. This was in fact the first thing I tested (against my own better judgement) :-)

**Comment by photopea** [comment]
No, I don't mean Base64 inside Data URI. I mean this: https://developer.mozilla.org/en-US/docs/Web/API/URL/createObjectURL

**Comment by titlestad** [comment]
Yes the blob API is great. But I still need to know how to pass it. A data url object is just a base64 encoded string, eg <img src="data:text/plain;base64,SGVsbG8sIFdvcmxkIQ%3D%3D"/> As you can see, the last part is encoded into the url and will become very large for a standard JPG image.

**Comment by titlestad** [comment]
Hmm, you're right. I will surely test this. I will come back to tell if it works in 5 minutes :-) - then we have the other one. What about save callbacks when I'm on a LAN behind a firewall?

**Comment by photopea** [comment]
I don't understand networks that much. I can advice you on the level of Web / HTTP, but the internet connection should be solved by others.

**Comment by titlestad** [comment]
Well, what I'm saying really is that a lot of our users are running Friend on a local server. So they aren't connected to a public server on the internet. This means that they don't have access to a php script which can receive your save callback.
Wouldn't it be easier to just receive the bytearray through postMessage? Then I could just receive the data with an onmessage() listener and store the data.

**Comment by photopea** [comment]
If they are not connected to the internet at all, they would not be able to load photopea.com inside an iframe.

I agree with you, there must be a way to pass files into and out the Photopea completely on the client. How much versatile should it be? wihh postMessage handle 1 GB files? Is there a way to avoid having two copies of the file?

**Comment by titlestad** [comment]
pp.js:5733 Failed to load blob:https://localhost:6502/3b7e60d6-b67a-4820-810a-59a9453b8d9d: Cross origin requests are only supported for protocol schemes: http, data, chrome, chrome-extension, https.

**Comment by titlestad** [comment]
No they're connected to the internet alright. Imagine a Friend PC, like a Chromebook. You start the kernel and log directly into the Friend Workspace on your PC. In this case, your server and everything is running on your local computer, but it's not connected to a public web server. So if the receiving end of the save loop is on your local computer, Photopea will not be able to contact it across the network. This is why a postmessage API would solve everything.

**Comment by titlestad** [comment]
Well, I don't think it's a huge problem. The blob can be a JPEG, so I don't think the message would be so large, it would be compressed. Postmessage is pretty fast.

**Comment by photopea** [comment]
Maybe we could remove the current "GET API" completely. Instead of having Photopea passing files to and from the server, it would only pass the file locally with the container, and the container would handle the rest.

**Comment by titlestad** [comment]
Yes I think that's a good idea. :-) I'll help test the API as much as you want. I really want Photopea to work "like a native app" in Friend. It would not only be a great killer app for us, but a great showroom for you as well. You could offer a full creative desktop in the cloud with Friend + Photopea. Friend is open source :-)

**Comment by photopea** [comment]
BTW. when you said you want to know about user pressing File - Open and show your Open dialog instead ... The current workflow expects, that you would hide the Open button in PP and have your own Open button outside Photopea. Once you have the file, you will pass it to PP.

**Comment by titlestad** [comment]
So what I'd like to do, Load:
1. Open file menu in Friend. Then the file will be read by the cloud server.
2. Our app containing the iframe would then post the binary array using the postmessage api.
3. Your app would receive the message with the data, unpacking the data and using it.

Save:
1. Save file in Friend menu.
2. Send postmessage to your api to retrieve the file, and format, perhaps callback id
3. Receive message from your api, with callback id and data blob. 
4. Friend unpacks the data and stores it to the cloud server.

Future: Think about the possibilities of having a web OS playing expert ball with your app :)

**Comment by titlestad** [comment]
Another thing, each time you select an active image in Pea, you could send a post message with "current file", so that my container could know which file would be saved. Etc.

**Comment by titlestad** [comment]
We could even write more filters if you'd allow to get buffer, and store new buffer. The possibilities are endless. We have a video conferencing solution in there too (Friend Chat) that would allow you to collaborate, share screens etc. So many ideas. :-)

**Comment by photopea** [comment]
Yes, that sounds truly amazing :) And the last bit is: get lots of people use it! :D  I will look at it the next week.

**Comment by titlestad** [comment]
Thanks! :-) Have a nice weekend man!

**Comment by photopea** [comment]
Hi, I extended the Live Messaging API. Now, you can send files to PP and get them from PP on a client side. See https://www.photopea.com/api/live  . What do you think about it?

**Comment by titlestad** [comment]
Fantastic! I will try it out right away!

**Comment by photopea** [comment]
@titlestad  So how did it go?

**Comment by SatishDash20** [comment]
I am trying to pass this image to photopea api for editing "http://www.converthink.com/public/images/analytics.png". How can i pass this image or how can i integrate photopea api into a project , will you please help me out, i am not getting proper idea from photopea/api please help me out.

**Comment by photopea** [comment]
Hi, you can try it on the Playground: https://www.photopea.com/api/playground , just rewrite the URL of the image with your own URL, and generate the URL of Photopea. 

Note, that your image has to be on HTTPS and the CORS has to be allowed. Learn how to use the web console, where errors are shown.
